File: ios/Scrapmate/UPIPaymentModule.swift
import Foundation
import UIKit
import React

@objc(UPIPaymentModule)
class UPIPaymentModule: NSObject, RCTBridgeModule {
    
    private var paymentPromise: RCTPromiseResolveBlock?
    private var paymentReject: RCTPromiseRejectBlock?
    
    static func moduleName() -> String! {
        return "UPIPaymentModule"
    }
    
    static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override init() {
        super.init()
        // Listen for payment response notifications
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handlePaymentResponseNotification(_:)),
            name: NSNotification.Name("UPIPaymentResponse"),
            object: nil
        )
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    @objc
    private func handlePaymentResponseNotification(_ notification: Notification) {
        if let userInfo = notification.userInfo,
           let response = userInfo["response"] as? String {
            handlePaymentResponse(response)
        }
    }
    
    @objc
    func initiatePayment(
        _ upiId: String,
        amount: String,
        transactionId: String,
        merchantName: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        DispatchQueue.main.async {
            self.paymentPromise = resolver
            self.paymentReject = rejecter
            
            // Create UPI payment URL
            let encodedUPIId = upiId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? upiId
            let encodedMerchantName = merchantName.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? merchantName
            let encodedTransactionNote = "Subscription Payment".addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "Subscription Payment"
            
            let upiURLString = "upi://pay?pa=\(encodedUPIId)&pn=\(encodedMerchantName)&am=\(amount)&tid=\(transactionId)&tr=\(transactionId)&tn=\(encodedTransactionNote)&cu=INR"
            
            guard let upiURL = URL(string: upiURLString) else {
                rejecter("INVALID_URL", "Invalid UPI URL", nil)
                return
            }
            
            // Check if any app can handle UPI URL
            if UIApplication.shared.canOpenURL(upiURL) {
                // Open UPI app
                UIApplication.shared.open(upiURL, options: [:]) { success in
                    if !success {
                        rejecter("OPEN_FAILED", "Failed to open UPI app", nil)
                    }
                    // Payment result will be handled via URL scheme callback in AppDelegate
                }
            } else {
                // Try alternative UPI apps
                let alternativeApps = [
                    "phonepe://pay",
                    "paytmmp://pay",
                    "gpay://pay",
                    "bhim://pay"
                ]
                
                var appOpened = false
                for appScheme in alternativeApps {
                    if let appURL = URL(string: appScheme),
                       UIApplication.shared.canOpenURL(appURL) {
                        UIApplication.shared.open(upiURL, options: [:]) { _ in }
                        appOpened = true
                        break
                    }
                }
                
                if !appOpened {
                    rejecter("NO_UPI_APP", "No UPI app found. Please install a UPI app like Google Pay, PhonePe, or Paytm.", nil)
                }
            }
        }
    }
    
    @objc
    func handlePaymentResponse(_ response: String) {
        // Parse UPI response
        let responseDict = parseUPIResponse(response)
        let status = responseDict["Status"] ?? "Unknown"
        
        if let resolver = paymentPromise {
            if status.uppercased() == "SUCCESS" {
                let result: [String: Any] = [
                    "status": "success",
                    "transactionId": responseDict["TxnId"] ?? "",
                    "responseCode": responseDict["ResponseCode"] ?? "",
                    "approvalRefNo": responseDict["ApprovalRefNo"] ?? ""
                ]
                resolver(result)
            } else {
                let result: [String: Any] = [
                    "status": "failed",
                    "message": responseDict["Status"] ?? "Payment failed"
                ]
                paymentReject?("PAYMENT_FAILED", "Payment failed", NSError(domain: "UPIPayment", code: 1, userInfo: result))
            }
            paymentPromise = nil
            paymentReject = nil
        }
    }
    
    private func parseUPIResponse(_ response: String) -> [String: String] {
        var dict: [String: String] = [:]
        let pairs = response.components(separatedBy: "&")
        for pair in pairs {
            let keyValue = pair.components(separatedBy: "=")
            if keyValue.count == 2 {
                let key = keyValue[0]
                let value = keyValue[1].removingPercentEncoding ?? keyValue[1]
                dict[key] = value
            }
        }
        return dict
    }
}

File: android/app/src/main/java/com/alpts/scrapmate/UPIPaymentModule.kt
package com.alpts.scrapmate

import android.app.Activity
import android.content.Intent
import android.net.Uri
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

class UPIPaymentModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext), ActivityEventListener {
    
    private var paymentPromise: Promise? = null
    
    init {
        reactContext.addActivityEventListener(this)
    }

    override fun getName(): String {
        return "UPIPaymentModule"
    }

    @ReactMethod
    fun initiatePayment(upiId: String, amount: String, transactionId: String, merchantName: String, promise: Promise) {
        try {
            // Store promise for later use
            paymentPromise = promise
            
            // Create UPI payment URI
            val uriString = buildString {
                append("upi://pay?")
                append("pa=").append(Uri.encode(upiId))
                append("&pn=").append(Uri.encode(merchantName))
                append("&am=").append(Uri.encode(amount))
                append("&tid=").append(Uri.encode(transactionId))
                append("&tr=").append(Uri.encode(transactionId))
                append("&tn=").append(Uri.encode("Subscription Payment"))
                append("&cu=INR")
            }
            
            val uri = Uri.parse(uriString)
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = uri
            
            // Check if any UPI app can handle this intent
            val chooser = Intent.createChooser(intent, "Pay with UPI")
            
            val activity = reactApplicationContext.currentActivity
            if (activity != null) {
                try {
                    activity.startActivityForResult(chooser, UPI_PAYMENT_REQUEST_CODE)
                } catch (e: Exception) {
                    promise.reject("NO_UPI_APP", "No UPI app found. Please install a UPI app like Google Pay, PhonePe, or Paytm.")
                }
            } else {
                promise.reject("NO_ACTIVITY", "Activity not available")
            }
        } catch (e: Exception) {
            promise.reject("PAYMENT_ERROR", e.message ?: "Failed to initiate payment")
        }
    }

    override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == UPI_PAYMENT_REQUEST_CODE) {
            paymentPromise?.let { promise ->
                when (resultCode) {
                    Activity.RESULT_OK -> {
                        // Payment might be successful, but we need to check the response
                        val response = data?.getStringExtra("response")
                        if (response != null) {
                            // Parse UPI response
                            val responseMap = parseUPIResponse(response)
                            val status = responseMap["Status"] ?: "Unknown"
                            
                            if (status.equals("SUCCESS", ignoreCase = true)) {
                                val result = Arguments.createMap().apply {
                                    putString("status", "success")
                                    putString("transactionId", responseMap["TxnId"])
                                    putString("responseCode", responseMap["ResponseCode"])
                                    putString("approvalRefNo", responseMap["ApprovalRefNo"])
                                }
                                promise.resolve(result)
                            } else {
                                val result = Arguments.createMap().apply {
                                    putString("status", "failed")
                                    putString("message", responseMap["Status"] ?: "Payment failed")
                                }
                                promise.reject("PAYMENT_FAILED", "Payment failed", result)
                            }
                        } else {
                            // No response data, check if payment was cancelled
                            val result = Arguments.createMap().apply {
                                putString("status", "cancelled")
                                putString("message", "Payment was cancelled")
                            }
                            promise.reject("PAYMENT_CANCELLED", "Payment was cancelled", result)
                        }
                    }
                    Activity.RESULT_CANCELED -> {
                        val result = Arguments.createMap().apply {
                            putString("status", "cancelled")
                            putString("message", "Payment was cancelled by user")
                        }
                        promise.reject("PAYMENT_CANCELLED", "Payment was cancelled", result)
                    }
                    else -> {
                        val result = Arguments.createMap().apply {
                            putString("status", "failed")
                            putString("message", "Payment failed")
                        }
                        promise.reject("PAYMENT_FAILED", "Payment failed", result)
                    }
                }
                paymentPromise = null
            }
        }
    }

    private fun parseUPIResponse(response: String): Map<String, String> {
        val map = mutableMapOf<String, String>()
        val pairs = response.split("&")
        for (pair in pairs) {
            val keyValue = pair.split("=", limit = 2)
            if (keyValue.size == 2) {
                map[keyValue[0]] = Uri.decode(keyValue[1])
            }
        }
        return map
    }

    override fun onNewIntent(intent: Intent) {
        // Handle deep link intents from UPI apps
        intent.data?.let { uri ->
            if (uri.scheme == "upi" || uri.scheme == "scrapmatepartner") {
                val response = uri.query ?: uri.toString()
                if (response.isNotEmpty()) {
                    handlePaymentResponse(response)
                }
            }
        }
    }
  
    // Public method to handle payment response from MainActivity
    fun handlePaymentResponse(response: String) {
        paymentPromise?.let { promise ->
            val responseMap = parseUPIResponse(response)
            val status = responseMap["Status"] ?: "Unknown"
            
            if (status.equals("SUCCESS", ignoreCase = true)) {
                val result = Arguments.createMap().apply {
                    putString("status", "success")
                    putString("transactionId", responseMap["TxnId"])
                    putString("responseCode", responseMap["ResponseCode"])
                    putString("approvalRefNo", responseMap["ApprovalRefNo"])
                }
                promise.resolve(result)
            } else {
                val result = Arguments.createMap().apply {
                    putString("status", "failed")
                    putString("message", responseMap["Status"] ?: "Payment failed")
                }
                promise.reject("PAYMENT_FAILED", "Payment failed", result)
            }
            paymentPromise = null
        }
    }

    companion object {
        private const val UPI_PAYMENT_REQUEST_CODE = 1001
    }
}

